File: paracci/tests/test_carrier_ui.py
import json
import re
from pathlib import Path

from jinja2 import Environment, FileSystemLoader


PACKAGE_ROOT = Path(__file__).resolve().parents[1]
TEMPLATE_ROOT = PACKAGE_ROOT / "app" / "templates"
STATIC_ROOT = PACKAGE_ROOT / "app" / "static"
I18N_ROOT = PACKAGE_ROOT / "app" / "i18n"

SETUP_TEMPLATE = (TEMPLATE_ROOT / "setup.html").read_text(encoding="utf-8")
SESSION_TEMPLATE = (TEMPLATE_ROOT / "session.html").read_text(encoding="utf-8")
BASE_TEMPLATE = (TEMPLATE_ROOT / "base.html").read_text(encoding="utf-8")
CARRIER_JS = (STATIC_ROOT / "js" / "carrier.js").read_text(encoding="utf-8")
SETUP_JS = (STATIC_ROOT / "js" / "setup.js").read_text(encoding="utf-8")
SESSION_JS = (STATIC_ROOT / "js" / "session.js").read_text(encoding="utf-8")
APP_JS = (STATIC_ROOT / "js" / "app.js").read_text(encoding="utf-8")
SESSION_CSS = (STATIC_ROOT / "css" / "pages" / "session.css").read_text(encoding="utf-8")

EXPECTED_CARRIER_KEYS = {
    "panel_title",
    "panel_optional",
    "outer_wrapper_note",
    "recompression_warning",
    "capacity_hint",
    "cover_file_label",
    "carrier_file_label",
    "choose_png",
    "no_png_selected",
    "select_png_error",
    "extract_setup",
    "extract_responder",
    "extract_message",
    "export_setup",
    "export_responder",
    "export_message",
    "insufficient_capacity",
    "generic_error",
    "processing",
}


def test_carrier_templates_compile():
    env = Environment(loader=FileSystemLoader(TEMPLATE_ROOT))
    env.filters["expires_at"] = lambda value: value
    env.filters["time_left"] = lambda value: value

    env.get_template("setup.html")
    env.get_template("session.html")


def test_carrier_controls_are_collapsed_secondary_and_explicit():
    combined = SETUP_TEMPLATE + SESSION_TEMPLATE

    assert "<details class=\"carrier-panel" in combined
    assert not re.search(r"<details[^>]*carrier-panel[^>]*\sopen(?:\s|=|>)", combined)
    assert 'class="carrier-panel carrier-option"' in SESSION_TEMPLATE
    assert 'class="carrier-panel mt-5"' not in SESSION_TEMPLATE
    assert ".carrier-option" in SESSION_CSS
    assert 'class="btn btn-secondary"' in combined
    assert "carrier.recompression_warning" in SETUP_TEMPLATE
    assert "carrier.recompression_warning" in SESSION_TEMPLATE
    assert "carrier.panel_optional" in combined
    assert "carrier.choose_png" in combined
    assert "carrier.no_png_selected" in combined
    assert "carrier-panel-badge" in combined
    assert "carrier-file-control" in combined

    for endpoint in (
        "main.session_import_carrier",
        "main.session_import_responder_carrier",
        "main.session_open_carrier",
        "main.session_export_carrier",
        "main.session_seal_carrier",
    ):
        assert endpoint in combined


def test_normal_paracci_flows_and_drop_targets_remain_primary():
    assert "url_for('main.session_import')" in SETUP_TEMPLATE
    assert 'name="paracci_file" id="fileInput" accept=".paracci" required' in SETUP_TEMPLATE
    assert 'data-drop-target="import"' in SETUP_TEMPLATE
    assert 'class="btn btn-primary btn-lg px-12" id="importBtn"' in SETUP_TEMPLATE

    assert "url_for('main.session_import_responder', sid=sid)" in SESSION_TEMPLATE
    assert 'name="paracci_file" id="responder-input" accept=".paracci"' in SESSION_TEMPLATE
    assert 'data-drop-target="finalize"' in SESSION_TEMPLATE
    assert "url_for('main.session_seal', sid=sid)" in SESSION_TEMPLATE
    assert 'id="seal-submit"' in SESSION_TEMPLATE
    assert 'id="paracci_file" name="paracci_file" accept=".paracci"' in SESSION_TEMPLATE
    assert 'data-drop-target="open"' in SESSION_TEMPLATE
    assert "url_for('main.session_export', sid=sid)" in SESSION_TEMPLATE


def test_post_bond_composer_state_is_rendered_hidden_and_backend_driven():
    assert 'id="bonded-checklist"{% if not meta.can_send %} hidden{% endif %}' in SESSION_TEMPLATE
    assert "{% if meta.role == 'Y' and not meta.can_send %}" in SESSION_TEMPLATE
    assert 'id="bond-pending-composer"' in SESSION_TEMPLATE
    assert 'id="message-composer"{% if not meta.can_send %} hidden{% endif %}' in SESSION_TEMPLATE
    pending_tag = re.search(r'<div id="bond-pending-composer"[^>]*>', SESSION_TEMPLATE)
    assert pending_tag
    assert "style=" not in pending_tag.group(0)
    assert 'class="bond-pending-composer"' in pending_tag.group(0)
    assert "shell_can_send = shell_meta.can_send" in BASE_TEMPLATE
    assert "shell_can_attach = endpoint == 'main.session_detail' and shell_can_send" in BASE_TEMPLATE
    assert "if (data?.session_can_send !== true) return;" in SESSION_JS
    assert "pendingComposer.remove();" in SESSION_JS
    assert "document.body.dataset.dropAttach = 'true';" in SESSION_JS
    assert SESSION_JS.count("applyPostOpenSessionState(data);") == 2
    assert "location.reload(" not in SESSION_JS


def test_carrier_inputs_are_explicit_png_pickers_without_path_or_drop_fields():
    combined = SETUP_TEMPLATE + SESSION_TEMPLATE + SETUP_JS + SESSION_JS

    assert 'accept="image/png,.png"' in SETUP_TEMPLATE
    assert 'accept="image/png,.png"' in SESSION_TEMPLATE
    assert "formData.set('carrier_png'" in SETUP_JS
    assert "formData.set('carrier_png'" in SESSION_JS
    assert "formData.set('cover_png'" in SESSION_JS
    assert "png_lossless_v1" in CARRIER_JS
    assert "carrier_native_file_id" not in combined
    assert "cover_native_file_id" not in combined
    assert 'name="carrier_png"' not in combined
    assert 'name="cover_png"' not in combined
    assert combined.count("data-carrier-drop-zone") >= 2
    assert "bindPngDropTarget" in SETUP_JS
    assert "bindPngDropTarget" in SESSION_JS

    carrier_input = re.search(
        r'<input type="file"[^>]+data-carrier-file[^>]*>',
        SESSION_TEMPLATE,
    )
    assert carrier_input
    assert "data-drop-target" not in carrier_input.group(0)
    for carrier_control in re.findall(
        r'<div class="carrier-file-control"[^>]*>',
        SETUP_TEMPLATE + SESSION_TEMPLATE,
    ):
        assert "data-carrier-drop-zone" in carrier_control
        assert "data-drop-target" not in carrier_control


def test_normal_drop_resolver_does_not_detect_png_carriers():
    resolver = re.search(
        r"function resolveDropIntent\(fileInfo = \{\}\) \{(?P<body>.*?)\n\}",
        APP_JS,
        re.DOTALL,
    )
    assert resolver
    body = resolver.group("body")
    assert ".paracci" in body
    assert ".png" not in body
    assert "carrier" not in body.lower()


def test_carrier_frontend_uses_safe_generic_errors_and_managed_downloads():
    combined = CARRIER_JS + SETUP_JS + SESSION_JS

    assert "alert.textContent = error?.kind === 'capacity'" in CARRIER_JS
    assert ": genericErrorText();" in CARRIER_JS
    assert "response.text(" not in CARRIER_JS
    assert "data.error" not in CARRIER_JS
    assert "console." not in CARRIER_JS
    assert "save_file_silent" in CARRIER_JS
    assert re.search(r"\.save_file\(", combined) is None
    for forbidden in ("FileReader", "readAsDataURL", "readAsArrayBuffer", "btoa(", "atob("):
        assert forbidden not in combined
    assert "qr_matrix_v1" not in combined
    assert "carrier_generic_error" in BASE_TEMPLATE
    assert "carrier_capacity_error" in BASE_TEMPLATE
    assert "carrier_select_png_error" in BASE_TEMPLATE
    assert "carrier_processing" in BASE_TEMPLATE
    assert "response.text(" not in CARRIER_JS
    assert "response.json(" in CARRIER_JS
    assert "data-carrier-file-name" in SETUP_TEMPLATE
    assert "data-carrier-file-name" in SESSION_TEMPLATE
    assert "output.textContent =" in CARRIER_JS


def test_all_locales_define_conservative_carrier_strings():
    locale_files = sorted(I18N_ROOT.glob("*.json"))
    assert {path.stem for path in locale_files} == {"de", "en", "es", "fr", "ru", "tr"}

    for locale_file in locale_files:
        payload = json.loads(locale_file.read_text(encoding="utf-8"))
        carrier = payload["carrier"]
        assert set(carrier) == EXPECTED_CARRIER_KEYS
        assert all(isinstance(value, str) and value.strip() for value in carrier.values())
        combined = " ".join(carrier.values()).lower()
        assert "undetect" not in combined
        assert "tamper-proof" not in combined
        assert "authenticated carrier" not in combined

    english = json.loads((I18N_ROOT / "en.json").read_text(encoding="utf-8"))["carrier"]
    assert "file/document" in english["recompression_warning"]
    assert "recompress" in english["recompression_warning"]
    assert "resize" in english["recompression_warning"]
    assert "convert" in english["recompression_warning"]
    assert "outer transport wrapper" in english["outer_wrapper_note"]
    assert "resolution" in english["capacity_hint"]
    assert "not PNG file size" in english["capacity_hint"]
    assert "attachments" in english["capacity_hint"].lower()
    assert "high-resolution lossless PNG" in english["capacity_hint"]
    assert "pixel capacity" in english["insufficient_capacity"]
    assert "attachments" in english["insufficient_capacity"].lower()
    assert "high-resolution lossless PNG" in english["insufficient_capacity"]
    assert "drop one here" in english["no_png_selected"].lower()
    assert "choose or drop a png file" in english["select_png_error"].lower()

    for locale_file in locale_files:
        carrier = json.loads(locale_file.read_text(encoding="utf-8"))["carrier"]
        for key in ("capacity_hint", "insufficient_capacity"):
            assert not re.search(
                r"\b\d+\s*(?:bytes?|kb|mb|gb|pixels?)\b|\b\d+\s*[x×]\s*\d+\b",
                carrier[key],
                re.IGNORECASE,
            )

File: paracci/tests/test_carrier_ui.mjs
import assert from 'node:assert/strict';
import fs from 'node:fs';
import path from 'node:path';
import test from 'node:test';
import vm from 'node:vm';
import { fileURLToPath } from 'node:url';

const TEST_DIR = path.dirname(fileURLToPath(import.meta.url));
const CARRIER_JS = fs.readFileSync(
    path.resolve(TEST_DIR, '../app/static/js/carrier.js'),
    'utf8'
);
const SETUP_JS = fs.readFileSync(
    path.resolve(TEST_DIR, '../app/static/js/setup.js'),
    'utf8'
);
const SESSION_JS = fs.readFileSync(
    path.resolve(TEST_DIR, '../app/static/js/session.js'),
    'utf8'
);
const APP_JS = fs.readFileSync(
    path.resolve(TEST_DIR, '../app/static/js/app.js'),
    'utf8'
);

function makeHarness({ fetchImpl, nativeApi = null } = {}) {
    const anchors = [];
    const objectUrls = [];
    const revokedUrls = [];
    const notices = [];

    const document = {
        body: {
            appendChild(node) {
                node.parentNode = this;
            }
        },
        createElement(tag) {
            if (tag === 'a') {
                const anchor = {
                    clicked: false,
                    removed: false,
                    click() {
                        this.clicked = true;
                    },
                    remove() {
                        this.removed = true;
                    }
                };
                anchors.push(anchor);
                return anchor;
            }
            return { className: '', textContent: '' };
        }
    };
    const TestURL = {
        createObjectURL(blob) {
            const value = `blob:carrier-${objectUrls.length + 1}`;
            objectUrls.push({ value, blob });
            return value;
        },
        revokeObjectURL(value) {
            revokedUrls.push(value);
        }
    };
    const window = {
        PARACCI_I18N: {
            carrier_generic_error: 'Localized generic carrier error.',
            carrier_capacity_error: 'Localized capacity error.',
            carrier_processing: 'Carrier busy.'
        },
        URL: TestURL,
        fetch: fetchImpl || (async () => {
            throw new Error('network sentinel');
        }),
        pywebview: nativeApi ? { api: nativeApi } : undefined,
        ParacciSecurity: {
            getLoopbackToken: () => 'loopback-token',
            navigateAuthorized: async () => true
        },
        showDownloadNotification(filename, savedPath) {
            notices.push({ filename, savedPath });
        }
    };
    const context = vm.createContext({
        Blob,
        document,
        window
    });
    vm.runInContext(CARRIER_JS, context, { filename: 'carrier.js' });

    return {
        anchors,
        api: window.ParacciCarrierUI,
        notices,
        objectUrls,
        revokedUrls,
        window
    };
}

test('generic carrier errors use localized text-safe rendering only', () => {
    const { api } = makeHarness();
    const children = [];
    const container = {
        replaceChildren() {
            children.length = 0;
        },
        appendChild(child) {
            children.push(child);
        }
    };

    api.showGenericError(container);

    assert.equal(children.length, 1);
    assert.equal(children[0].className, 'alert alert-error');
    assert.equal(children[0].textContent, 'Localized generic carrier error.');
});

test('network and backend failures do not expose sentinel details', async () => {
    const network = makeHarness({
        fetchImpl: async () => {
            throw new Error('payload-token-path-sentinel');
        }
    });
    await assert.rejects(
        network.api.submitOpen('/session/id/carrier/open', {}),
        error => {
            assert.equal(error.message, 'Localized generic carrier error.');
            assert.doesNotMatch(error.message, /payload|token|path|sentinel/);
            return true;
        }
    );

    let jsonRead = false;
    const backend = makeHarness({
        fetchImpl: async () => ({
            ok: false,
            async json() {
                jsonRead = true;
                return { error: 'backend-magic-checksum-sentinel' };
            }
        })
    });
    await assert.rejects(
        backend.api.submitOpen('/session/id/carrier/open', {}),
        /Localized generic carrier error/
    );
    assert.equal(jsonRead, false);
});

test('carrier imports follow only authenticated redirect navigation', async () => {
    let navigatedTo = '';
    const harness = makeHarness({
        fetchImpl: async () => ({
            ok: true,
            redirected: true,
            url: 'http://127.0.0.1/session/opaque'
        })
    });
    harness.window.ParacciSecurity.navigateAuthorized = async url => {
        navigatedTo = url;
        return true;
    };

    await harness.api.submitImport('/session/import/carrier', {});

    assert.equal(navigatedTo, 'http://127.0.0.1/session/opaque');
});

test('browser carrier output uses a blob URL download', async () => {
    const blob = new Blob(['png']);
    const harness = makeHarness({
        fetchImpl: async () => ({
            ok: true,
            headers: {
                get(name) {
                    return name === 'Content-Disposition'
                        ? 'attachment; filename="holiday-photo-carrier.png"'
                        : null;
                }
            },
            async blob() {
                return blob;
            }
        })
    });

    await harness.api.submitDownload('/session/id/carrier/export', {});

    assert.equal(harness.anchors.length, 1);
    assert.equal(harness.anchors[0].download, 'holiday-photo-carrier.png');
    assert.equal(harness.anchors[0].clicked, true);
    assert.equal(harness.anchors[0].removed, true);
    assert.equal(harness.objectUrls[0].blob, blob);
    assert.deepEqual(harness.revokedUrls, [harness.objectUrls[0].value]);
});

test('native carrier output consumes the existing one-shot save grant', async () => {
    const calls = [];
    const nativeApi = {
        async save_file_silent(token, loopbackToken) {
            calls.push({ token, loopbackToken });
            return 'managed-output';
        }
    };
    let nativeHeader = '';
    const harness = makeHarness({
        nativeApi,
        fetchImpl: async (_url, init) => {
            nativeHeader = init.headers['X-Paracci-Native-Save'];
            return {
                ok: true,
                async json() {
                    return {
                        native_save_token: 'one-shot-grant',
                        filename: 'holiday-photo-carrier.png'
                    };
                }
            };
        }
    });

    await harness.api.submitDownload('/session/id/carrier/export', {});

    assert.equal(nativeHeader, '1');
    assert.deepEqual(calls, [{
        token: 'one-shot-grant',
        loopbackToken: 'loopback-token'
    }]);
    assert.deepEqual(harness.notices, [{
        filename: 'holiday-photo-carrier.png',
        savedPath: 'managed-output'
    }]);
});

test('capacity failures use localized safe text without reading backend details', async () => {
    let jsonRead = false;
    let textRead = false;
    const harness = makeHarness({
        fetchImpl: async () => ({
            ok: false,
            status: 422,
            async json() {
                jsonRead = true;
                return { error: 'capacity-byte-count-path-sentinel' };
            },
            async text() {
                textRead = true;
                return 'capacity-byte-count-path-sentinel';
            }
        })
    });

    await assert.rejects(
        harness.api.submitDownload('/session/id/carrier/seal', {}),
        error => {
            assert.equal(error.kind, 'capacity');
            assert.equal(error.message, 'Localized capacity error.');
            assert.doesNotMatch(error.message, /byte|count|path|sentinel/);
            return true;
        }
    );
    assert.equal(jsonRead, false);
    assert.equal(textRead, false);
});

test('carrier filenames reject path-like or unsafe response values', () => {
    const { api } = makeHarness();

    assert.equal(
        api.safePngFilename('attachment; filename="holiday-photo-carrier.png"'),
        'holiday-photo-carrier.png'
    );
    assert.equal(api.safePngFilename('../../private.png'), 'image-carrier.png');
    assert.equal(api.safePngFilename('token sentinel.txt'), 'image-carrier.png');
});

test('selected carrier filenames render through textContent', () => {
    const { api } = makeHarness();
    const output = {
        dataset: { emptyLabel: 'No PNG selected' },
        textContent: ''
    };
    const input = {
        files: [{ name: '<img src=x onerror=token-sentinel>.png' }],
        closest() {
            return {
                querySelector() {
                    return output;
                }
            };
        }
    };

    api.updateSelectedFileName(input);

    assert.equal(output.textContent, '<img src=x onerror=token-sentinel>.png');
    assert.equal('innerHTML' in output, false);
});

test('page wiring uses explicit multipart carrier fields only', () => {
    assert.match(SETUP_JS, /formData\.delete\('paracci_file'\)/);
    assert.match(SETUP_JS, /formData\.delete\('native_file_id'\)/);
    assert.match(SETUP_JS, /formData\.set\('carrier_png'/);
    assert.match(SESSION_JS, /formData\.set\('carrier_png'/);
    assert.match(SESSION_JS, /formData\.set\('cover_png'/);
    assert.match(SESSION_JS, /new FormData\(sourceForm\)/);
    for (const requiredField of [
        'message',
        'ttl_seconds',
        'allow_download',
        'attachments',
        'staged_attachment_ids'
    ]) {
        assert.doesNotMatch(SESSION_JS, new RegExp(`formData\\.delete\\('${requiredField}'\\)`));
    }
    assert.match(CARRIER_JS, /png_lossless_v1/);

    const combined = CARRIER_JS + SETUP_JS + SESSION_JS;
    for (const forbidden of [
        'carrier_native_file_id',
        'cover_native_file_id',
        'FileReader',
        'readAsDataURL',
        'readAsArrayBuffer',
        'btoa(',
        'atob(',
        'qr_matrix_v1'
    ]) {
        assert.doesNotMatch(combined, new RegExp(forbidden.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')));
    }
    assert.doesNotMatch(combined, /\.save_file\(/);
});

test('normal drop intent remains limited to paracci files and attachments', () => {
    const match = APP_JS.match(
        /function resolveDropIntent\(fileInfo = \{\}\) \{([\s\S]*?)\n\}/
    );
    assert.ok(match);
    assert.match(match[1], /\.paracci/);
    assert.doesNotMatch(match[1], /\.png|carrier/i);
});
